File: code/functions/function-type.m
#import <Foundation/Foundation.h>

// Function types in Objective-C
typedef int (*FuncType)(int, int);

int add(int a, int b) {
    return a + b;
}

int multiply(int a, int b) {
    return a * b;
}

int applyOperation(FuncType op, int x, int y) {
    return op(x, y);
}

int main() {
    FuncType operation = add;
    int result = applyOperation(operation, 5, 3);
    NSLog(@"Result: %d", result);
    
    operation = multiply;
    result = applyOperation(operation, 5, 3);
    NSLog(@"Result: %d", result);
    
    return 0;
}

File: code/functions/function-type.erl
% Function types in Erlang
-module(example).
-export([add/2, multiply/2, apply_operation/3, start/0]).

add(A, B) ->
    A + B.

multiply(A, B) ->
    A * B.

apply_operation(Op, X, Y) ->
    Op(X, Y).

start() ->
    Operation = fun add/2,
    Result = apply_operation(Operation, 5, 3),
    io:format("Result: ~p~n", [Result]),
    
    Operation2 = fun multiply/2,
    Result2 = apply_operation(Operation2, 5, 3),
    io:format("Result: ~p~n", [Result2]).